runtime/commonMain/src/kotlinx/serialization/internal/PrimitiveArraysSerializers.kt

```diff
@@ -0,0 +1,398 @@
+/*
+ * Copyright 2017-2019 JetBrains s.r.o. Use of this source code is governed by the Apache 2.0 license.
+ */
+
+package kotlinx.serialization.internal
+
+import kotlinx.serialization.*
+
+private const val INITIAL_SIZE = 0
+
+// it could be an object, but https://youtrack.jetbrains.com/issue/KT-14628
+/**
+ * Serializer for [ByteArray].
+ *
+ * Encode elements one-by-one, as regular list,
+ * unless format's Encoder/Decoder have special handling for this serializer.
+ */
+public class ByteArraySerializer :
+    PrimitiveArraySerializer<Byte, ByteArray, ByteArraySerializer.Builder>(ByteSerializer, ByteDescriptor),
+    KSerializer<ByteArray> {
+    inner class Builder internal constructor() :
+        PrimitiveArraySerializer<Byte, ByteArray, ByteArraySerializer.Builder>.Builder() {
+
+        private var buf: ByteArray = ByteArray(INITIAL_SIZE)
+        override var position: Int = 0
+            private set
+
+        constructor(filledBuf: ByteArray): this() {
+            buf = filledBuf
+            position = filledBuf.size
+        }
+
+        override fun ensureCapacity(requiredCapacity: Int) {
```

Comment:
This method should check that provided capacity is >= -1 if invoked from other classes and shouldn't if invoked from `append`

Reply:
Buffer size is not negative, so `buf.size < requiredCapacity` is always false on capacity < 0. Why do we need an additional check?

Reply:
Because that's clearly a programmer error (e.g. off-by-one error).
And the best action here is to notify the programmer about such error, not to ignore it.